juju/model.py

```diff
@@ -0,0 +1,172 @@
+            raise TooManyRelatedApps(relation_name, num_related, 1)
+
+
+class ModelEntityCache(WeakValueDictionary):
+    def __init__(self, local_unit_name):
+        super().__init__()
+        self.local_unit_name = local_unit_name
+
+    def get(self, entity_type, entity_name):
+        key = (entity_type, entity_name)
+        entity = super().get(key)
+        if entity is None:
+            entity = entity_type(entity_name, self)
+            self[key] = entity
+        return entity
+
+
+class Application:
+    def __init__(self, name, entity_cache):
+        self.name = name
+
+    def __repr__(self):
+        return f'<{type(self).__module__}.{type(self).__name__} {self.name}>'
+
+
+class Unit:
+    def __init__(self, name, entity_cache):
+        self.name = name
+        self.app = entity_cache.get(Application, name.split('/')[0])
+
+    def __repr__(self):
+        return f'<{type(self).__module__}.{type(self).__name__} {self.name}>'
+
+
+class LazyMapping(Mapping, ABC):
+    @abstractmethod
+    def _load(self):
+        raise NotImplementedError()
+
+    @property
+    def _data(self):
+        if not hasattr(self, '_lazy_data'):
+            self._lazy_data = self._load()
+        return self._lazy_data
+
+    def __contains__(self, key):
+        return key in self._data
+
+    def __len__(self):
+        return len(self._data)
+
+    def __iter__(self):
+        return iter(self._data)
+
+    def __getitem__(self, key):
+        return self._data[key]
+
+
+class RelationMap(LazyMapping):
+    """Map of relation names to lists of Relation instances."""
+    def __init__(self, relation_names, backend, entity_cache):
+        self._relation_names = relation_names
+        self._backend = backend
+        self._entity_cache = entity_cache
+
+    def _load(self):
+        data = {}
+        for relation_name in self._relation_names:
+            relations = data[relation_name] = []
+            for relation_id in self._backend.relation_ids(relation_name):
+                relations.append(Relation(relation_name, relation_id, self._backend, self._entity_cache))
+        return data
+
+
+class Relation:
+    def __init__(self, relation_name, relation_id, backend, entity_cache):
+        self.relation_name = relation_name
+        self.relation_id = relation_id
+        self._backend = backend
+        self.data = RelationData(self.relation_name, relation_id, self._backend, entity_cache)
+
+    @property
+    def apps(self):
+        return [unit.app for unit in self.units]
+
+    @property
+    def units(self):
+        return list(self.data.keys())
+
+
+class RelationData(LazyMapping):
+    def __init__(self, relation_name, relation_id, backend, entity_cache):
+        self.relation_name = relation_name
+        self.relation_id = relation_id
+        self._backend = backend
+        self._entity_cache = entity_cache
+
+    def _load(self):
+        data = {}
+        for unit_name in self._backend.relation_list(self.relation_id):
```

Comment:
Instead of listing units for the relation here, I think we need that `.units` field in the `Relation` instance, and provide that upfront. This way we have a list of units in the relation that will certainly be useful, and we don't need to call out to the backend and external command here.